NitroxClient/MonoBehaviours/Debugging/NetworkDebugger.cs

```diff
@@ -0,0 +1,36 @@
+﻿using UnityEngine;
+
+namespace NitroxClient.MonoBehaviours.Debugging
+{
+    public class NetworkDebugger : BaseDebugger
+    {
+        public void Awake()
+        {
+            Hotkey = KeyCode.N;
```

Comment:
It looks like these fields are only initialized once, are not supposed to be changed thereafter, and do not need to be set specifically from `Awake`. How about moving their initialisation to a constructor, and specifying the fields as `readonly`? More specifically, since the fields are mandatory, require them to be passed as arguments to a constructor in `BaseDebugger`?

Reply:
public NetworkDebugger() : base(400, KeyCode.N, true, false, false, GUISkinCreationOptions.DERIVEDCOPY)

Doesn't work because Unity won't call the `base()` constructor because classes implement `MonoBehaviour` and act like 'scripts'. Has to do with how Unity works. I can't use `readonly` either because I can't set it from the derived constructor (in SceneDebugger/NetworkDebugger).

I could refactor as much code as possible into a normal C# class for both debuggers but it won't be pretty in the projects hierarchy imo. I can break convention or work around Unity. Which would you say is best?

Reply:
Of course, I forgot Unity instantiates objects similarly to deserializing; without calling the constructor.

Maybe use get-properties?

Reply:
I've had success in the past by building a "core" library containing models that encapsulated all of my behaviors. The public API's were structured with Unity in mind, but each and every one was a SOLID object.

My MonoBehavior objects then take on an Adapter pattern as they are really only responsible for communicating user interactions between Unity and my game's real engine. I initialize an instance of a model that implements a specific behavior during the Awake of a MonoBehavior. Then, it is just a matter of calling the appropriate engine API method whenever an action is invoked on the MonoBehavior.

I don't know what the rules are for linking to other repos, but I do have a very simple example that demonstrates the design pattern I am referring to. If you're interested, then I think that could lend itself towards solving the problem of encapsulation while also dealing with the procedural nature Unity itself.